### PR TITLE
worklets plugin: preserve JSX imports in bundle mode

### PR DESCRIPTION
**Pull request by hannojg**
## Summary

This adds support for importing JSX statements/imports in worklets bundle mode!

## Test plan

- Unit tests
- example app babel plugin tests:

Screenshot 2026-06-06 at 12 56 56

### CODE REVIEW

**Review by tjzel — COMMENTED** [review]
As we discussed briefly off-discord - I think there's no need to add a separate option for that. In Bundle Mode we should preserve JSX as long as it comes from modules allowed for imports.

Could you also add a runtime tests suite in `/apps/common-app/src/apps/reanimated/examples/RuntimeTests/tests/plugin` that checks if worklets with JSX work at runtime?

**Comment by hannojg** [comment]
Hey @tjzel, i reworked the PR to just preserve JSX in general, without hiding it between an option.

When testing in the example app i ran into this issue though:
- https://github.com/software-mansion/react-native-reanimated/issues/9214

I made a "fix" for that here:
- https://github.com/software-mansion/react-native-reanimated/pull/9610

This PR depends on the mentioned PR.

**Comment by tjzel** [comment]
@hannojg I opened a separate PR that seems to be a general fix for multiple passes of generated files #9618, could you see if it can replace your other one?

**Comment by hannojg** [comment]
I checked this against latest `main` with #9618 included.

I made a [test branch and applied only the JSX capture changes from this PR](https://github.com/hannojg/react-native-reanimated/pull/new/hannojg/test-jsx-capture-on-latest-main), without the generated JSX transform fix from #9610. The unit tests and typecheck pass, but bundling the Fabric example in bundle mode still fails with the same JSX dev transform error:

```text
Duplicate __self prop found
packages/react-native-worklets/.worklets/17194764026949.js
return <ImportedComponent __self={this} />;
```

So #9618 fixes the multiple-pass/autoworkletization case, but it does not seem to cover this specific JSX-in-generated-worklet-file case. So it seems like something like #9610 is still required unfortunately :/

File: packages/react-native-worklets/plugin/src/generate.ts [hunk 1]
```diff
@@ -69,9 +69,15 @@ export function generateWorkletFile(
...
+  const reactPreset = require.resolve('@babel/preset-react');
+  const reactPresetOptions = {
+    development: false,
+    runtime: 'automatic',
+  };
+
   const transformedProg = transformFromAstSync(newProg, undefined, {
     filename: state.file.opts.filename,
-    presets: ['@babel/preset-typescript'],
+    presets: ['@babel/preset-typescript', [reactPreset, reactPresetOptions]],
```
Comment (tjzel) [inline packages/react-native-worklets/plugin/src/generate.ts]: @hannojg 

I looked into it and it turns out that the transformation that adds `__self` is added when entering the whole Program so when generating the worklet file we already have that in.

I think that instead of transforming it with react-preset to `_jsx` form maybe we should revert the work of that plugin, namely strip the `__self` and `__source` props to be closer to a real source file https://babeljs.io/docs/babel-preset-react#development.

A minimal plugin like this:

```diff
diff --git a/packages/react-native-worklets/plugin/src/closure.ts b/packages/react-native-worklets/plugin/src/closure.ts
index bac569c82d..34e998b910 100644
--- a/packages/react-native-worklets/plugin/src/closure.ts
+++ b/packages/react-native-worklets/plugin/src/closure.ts
@@ -32,7 +32,7 @@ export function getClosure(
         typePath.skip();
       },
       ReferencedIdentifier(idPath) {
-        if (idPath.isJSXIdentifier()) {
+        if (idPath.isJSXIdentifier() && !state.opts.bundleMode) {
           return;
         }
 
diff --git a/packages/react-native-worklets/plugin/src/generate.ts b/packages/react-native-worklets/plugin/src/generate.ts
index 32906d4244..87a54d9d26 100644
--- a/packages/react-native-worklets/plugin/src/generate.ts
+++ b/packages/react-native-worklets/plugin/src/generate.ts
@@ -4,6 +4,7 @@ import type {
   FunctionExpression,
   ImportDeclaration,
   ImportSpecifier,
+  JSXAttribute,
 } from '@babel/types';
 import {
   cloneNode,
@@ -72,7 +73,7 @@ export function generateWorkletFile(
   const transformedProg = transformFromAstSync(newProg, undefined, {
     filename: state.file.opts.filename,
     presets: ['@babel/preset-typescript'],
-    plugins: [state.autoworkletizationPlugin],
+    plugins: [state.autoworkletizationPlugin, stripJsxDevAttributesPlugin],
     ast: false,
     babelrc: false,
     configFile: false,
@@ -85,3 +86,18 @@ export function generateWorkletFile(
 
   writeFileSync(dedicatedFilePath, transformedProg);
 }
+
+const stripJsxDevAttributesPlugin = {
+  name: 'worklets-strip-jsx-dev-attributes',
+  visitor: {
+    JSXAttribute(path: NodePath<JSXAttribute>) {
+      const name = path.node.name;
+      if (
+        name.type === 'JSXIdentifier' &&
+        (name.name === '__self' || name.name === '__source')
+      ) {
+        path.remove();
+      }
+    },
+  },
+};
```

solved the issue for me with metro bundling. Could you see if it works for you too?

Comment (hannojg) [reply packages/react-native-worklets/plugin/src/generate.ts]: sorry for the late reply, yes i tested and it actually works well! Do you want me to incorporate that in this PR, or is that rather something you want to land yourself?

Comment (tjzel) [reply packages/react-native-worklets/plugin/src/generate.ts]: We plan to release 0.10.0 in a few days, so if lend me a hand, incorporate that and amend the tests, it would be very helpful!

**Comment by hannojg** [comment]
okay, added in your suggested approach!

File: apps/common-app/src/apps/reanimated/examples/RuntimeTests/tests/plugin/jsxInWorklets.test.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            output.value = typeof element.type === 'function';
+          });
+        });
+
+        return null;
+      };
+
+      await render(<ExampleComponent />);
+      await wait(100);
+      const sharedValue = await getRegisteredValue(SHARED_VALUE_REF);
```
Comment (tjzel) [inline apps/common-app/src/apps/reanimated/examples/RuntimeTests/tests/plugin/jsxInWorklets.test.tsx]: shared value syntax for tests is kind-of outdated, please re-write it to a less obfuscated approach used in i.e. tests for `scheduleOnUI`

File: packages/react-native-worklets/__tests__/plugin.test.ts
Comment (tjzel) [inline packages/react-native-worklets/__tests__/plugin.test.ts:1]: plugin.test.ts is for bundle mode disabled (it should've been added in the name, sorry for that), so there's no need to add bundleMode tests here

File: packages/react-native-worklets/__tests__/plugin.test.ts [hunk 3]
```diff
@@ -61,6 +67,49 @@ function runPlugin(
...
+    cwd: process.cwd(),
+    filename: __filename,
+    workletNumber: 0,
+    classesToWorkletize: [],
+  } as WorkletsPluginPass;
+
+  initializeState(state);
+
+  return getClosure(workletPath, state);
+}
```
Comment (tjzel) [inline packages/react-native-worklets/__tests__/plugin.test.ts]: We use some local string matchers to check closure (outside of bundle mode) in other tests, no need to add that helper.

File: packages/react-native-worklets/plugin/src/closure.ts [hunk 4]
```diff
@@ -32,7 +36,7 @@ export function getClosure(
         typePath.skip();
       },
       ReferencedIdentifier(idPath) {
-        if (idPath.isJSXIdentifier()) {
+        if (shouldSkipJSXIdentifier(idPath, state)) {
```
Comment (tjzel) [inline packages/react-native-worklets/plugin/src/closure.ts]: Why has it diverged from the diff? I think we should only bail here if bundle mode is disabled. Are there some other edge cases which we must cover with a more complicated check?

File: packages/react-native-worklets/__tests__/plugin-bundleMode.test.ts [hunk 5]
```diff
@@ -200,6 +205,44 @@ describe('babel plugin in bundleMode', () => {
...
+        autoworkletizationPlugin,
+      } as WorkletsPluginPass;
+
+      generateWorkletFile(new Set(), new Set(), factory, 1, state);
+      const files = capturedFiles;
+      expect(files).toHaveLength(1);
+      expect(files[0].content).toContain('return <ImportedComponent />;');
+      expect(files[0].content).not.toContain('__self');
+      expect(files[0].content).not.toContain('__source');
+    });
```
Comment (tjzel) [inline packages/react-native-worklets/__tests__/plugin-bundleMode.test.ts:242]: let's actually test it end-to-end, without deep imports

**Comment by hannojg** [comment]
thanks for the thorough review, i tried to apply everything! wdyt?

**Comment by tjzel** [comment]
@hannojg All looks good, I'm just wondering why is that additional check required here?

https://github.com/software-mansion/react-native-reanimated/pull/9212/changes#diff-ccfd9e7903007801372efe098b98035474cafbedcd42fbba90cc9b19d0d62787R80-R82

**Comment by tjzel** [comment]
I took the liberty to make the unit tests a little more robust.

**Comment by hannojg** [comment]
sorry which additional checks? i feel like you referenced either outdated or the wrong code? 👀

**Review by tjzel — APPROVED** [review]
